cfg/CFG.cc

```diff
@@ -22,97 +32,161 @@ BasicBlock *CFG::freshBlock(int outerLoops, int rubyBlockId) {
             }
 
             if (auto *v = cast_instruction<Ident>(bind.value.get())) {
-                blockReads.insert(v->what);
-                blockReadsAndWrites.insert(v->what);
+                blockReads.insert(v->what.id());
             } else if (auto *v = cast_instruction<Send>(bind.value.get())) {
-                blockReads.insert(v->recv.variable);
-                blockReadsAndWrites.insert(v->recv.variable);
+                blockReads.insert(v->recv.variable.id());
                 for (auto &arg : v->args) {
-                    blockReads.insert(arg.variable);
-                    blockReadsAndWrites.insert(arg.variable);
+                    blockReads.insert(arg.variable.id());
                 }
             } else if (auto *v = cast_instruction<TAbsurd>(bind.value.get())) {
-                blockReads.insert(v->what.variable);
+                blockReads.insert(v->what.variable.id());
             } else if (auto *v = cast_instruction<Return>(bind.value.get())) {
-                blockReads.insert(v->what.variable);
-                blockReadsAndWrites.insert(v->what.variable);
+                blockReads.insert(v->what.variable.id());
             } else if (auto *v = cast_instruction<BlockReturn>(bind.value.get())) {
-                blockReads.insert(v->what.variable);
-                blockReadsAndWrites.insert(v->what.variable);
+                blockReads.insert(v->what.variable.id());
             } else if (auto *v = cast_instruction<Cast>(bind.value.get())) {
-                blockReads.insert(v->value.variable);
-                blockReadsAndWrites.insert(v->value.variable);
+                blockReads.insert(v->value.variable.id());
             } else if (auto *v = cast_instruction<LoadSelf>(bind.value.get())) {
-                blockReads.insert(v->fallback);
-                blockReadsAndWrites.insert(v->fallback);
+                blockReads.insert(v->fallback.id());
             } else if (auto *v = cast_instruction<SolveConstraint>(bind.value.get())) {
-                blockReads.insert(v->send);
-                blockReadsAndWrites.insert(v->send);
+                blockReads.insert(v->send.id());
             }
 
-            auto fnd = blockReads.find(bind.bind.variable);
-            if (fnd == blockReads.end()) {
-                blockDead.insert(bind.bind.variable);
+            if (!blockReads.contains(bind.bind.variable.id())) {
+                blockDead.insert(bind.bind.variable.id());
             }
         }
         ENFORCE(bb->bexit.cond.variable.exists());
-        if (bb->bexit.cond.variable != core::LocalVariable::unconditional()) {
-            blockReads.insert(bb->bexit.cond.variable);
-            blockReadsAndWrites.insert(bb->bexit.cond.variable);
+        if (bb->bexit.cond.variable != LocalRef::unconditional()) {
+            blockReads.insert(bb->bexit.cond.variable.id());
         }
+
+        // Convert sets to sorted vectors.
```

Comment:
For large input sets, this up-front conversion cost pays huge dividends later when we take the union of sets / iterate through the entries.